src/coreclr/src/nativeaot/Bootstrap/main.cpp

```diff
@@ -101,253 +94,18 @@ static char& __unbox_z = __stop___unbox;
-extern "C" __NORETURN void RhpThrowEx(void * pEx);
-extern "C" void RhDebugBreak();
-
-extern "C" Object * __allocate_object(MethodTable * pMT)
-{
-    return RhNewObject(pMT);
-}
-
-extern "C" Object * __allocate_array(size_t elements, MethodTable * pMT)
-{
-    return RhNewArray(pMT, (int32_t)elements); // TODO: type mismatch
-}
-
-extern "C" Object * __castclass(MethodTable * pTargetMT, void* obj)
-{
-    return (Object *)RhTypeCast_CheckCast(pTargetMT, obj);
-}
-
-extern "C" Object * __isinst(MethodTable * pTargetMT, void* obj)
-{
-    return (Object *)RhTypeCast_IsInstanceOf(pTargetMT, obj);
-}
-
-extern "C" void __stelem_ref(void * pArray, unsigned idx, void * obj)
-{
-    RhpStelemRef(pArray, idx, obj);
-}
-
-extern "C" void* __ldelema_ref(void * pArray, unsigned idx, MethodTable * type)
-{
-    return RhpLdelemaRef(pArray, idx, type);
-}
-
-extern "C" void __throw_exception(void * pEx)
-{
-    RhpThrowEx(pEx);
-}
-
-extern "C" void __debug_break()
-{
-    RhDebugBreak();
-}
-
-void __range_check_fail()
-{
-    throw "ThrowRangeOverflowException";
-}
-
-extern "C" void RhpReversePInvoke2(ReversePInvokeFrame* pRevFrame);
-extern "C" void RhpReversePInvokeReturn2(ReversePInvokeFrame* pRevFrame);
-
-void __reverse_pinvoke(ReversePInvokeFrame* pRevFrame)
-{
-    RhpReversePInvoke2(pRevFrame);
-}
-
-void __reverse_pinvoke_return(ReversePInvokeFrame* pRevFrame)
-{
-    RhpReversePInvokeReturn2(pRevFrame);
-}
-
-extern "C" void RhpPInvoke2(PInvokeTransitionFrame* pFrame);
-extern "C" void RhpPInvokeReturn2(PInvokeTransitionFrame* pFrame);
-
-void __pinvoke(PInvokeTransitionFrame* pFrame)
-{
-    RhpPInvoke2(pFrame);
-}
-
-void __pinvoke_return(PInvokeTransitionFrame* pFrame)
-{
-    RhpPInvokeReturn2(pFrame);
-}
-
-namespace System_Private_CoreLib { namespace System { 
-
-    class Object {
-    public:
-        MethodTable * get_EEType() { return *(MethodTable **)this; }
-    };
-
-    class Array : public Object {
-    public:
-        int32_t GetArrayLength() {
-            return *(int32_t *)((void **)this + 1);
-        }
-        void * GetArrayData() {
-            return (void **)this + 2;
-        }
-    };
-
-    class String : public Object { public:
-        static MethodTable * __getMethodTable();
-    };
-
-    class String__Array : public Object { public:
-        static MethodTable * __getMethodTable();
-    };
-
-    class EETypePtr { public:
-        intptr_t m_value;
-    };
-
-}; };
-
-Object * __load_string_literal(const char * string)
-{
-    // TODO: Cache/intern string literals
-    // TODO: Unicode string literals
-
-    size_t len = strlen(string);
-
-    Object * pString = RhNewArray(System_Private_CoreLib::System::String::__getMethodTable(), (int32_t)len);
-
-    uint16_t * p = (uint16_t *)((char*)pString + sizeof(intptr_t) + sizeof(int32_t));
-    for (size_t i = 0; i < len; i++)
-        p[i] = string[i];
-    return pString;
-}
-
-#if defined(HOST_WASM)
```

Comment:
I never realized that, but WASM builds against bootstrappercpp.a, so these CPPCODEGEN ifdefs also apply to WASM :(.

We'll need to decouple that to keep HOST_WASM.

Reply:
I think the approach we should take is to delete this cruft as we run into it in NativeAOT branch, and then rewind it as necessary in the eventual wasm branch.